eclipse-che/che-operator
# feat: Import Devworkspace Che Operator

### What does this PR do?
This imports the Devworkspace Che Operator as a new controller inside `che-operator`.

The minimum amount of changes have been made to both codebases for the stuff to work.
In another words, the new controller `pkg/controller/devworkspace` is essentially still
completely standalone. It even still runs in its own reconcile loop even though it
reconciles `CheCluster` which is also reconciled by the main `che` controller.

The significant changes are are concentrated in `cmd/manager/main.go`, which sets the whole
show up.

In `pkg/deploy/dev-workspace`, we have a new approach for installing the devworkspace support.
The operator exits after a successfull installation, relying on Kubernetes to restart it.
When intialized again, it sees the CRDs for devworkspaces present in the cluster and
initializes the devworkspace reconciler.

### Screenshot/screencast of this PR
N/A

### What issues does this PR fix or reference?
eclipse/che#19408

### How to test this PR?
1. Install che-operator
1. Remove the DWCO container for the che-operator deployment (not yet done in this PR)
1. Change the che-operator image to `quay.io/lkrejci/che-operator:issue-19408`
1. Enable devworkspaces in the CheCluster CR
1. Notice the che-operator restart automatically after installation of Devworkspaces
1. Try to create a devfile v2.
1. Rejoice.

### PR Checklist
          
[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
